### DISCUSSION TITLE
Generate types similar to RecipeVariants, but with string literals

### DISCUSSION THREAD

**Discussion by laura-asimov** [vanilla-extract-css/vanilla-extract#1089] [Q&A] [answered]
I'm looking for a way to generate the type from a `recipe` and have it include only the string literals
I currently generate the type using `export type TextVariants = RecipeVariants<typeof textRecipe>;`

<img width="569" alt="Screenshot 2023-05-12 at 15 14 34" src="https://github.com/vanilla-extract-css/vanilla-extract/assets/111299974/20263e51-c2cc-4ce7-b3df-df9065751787">

In this scenario I'm looking to export the type as
`export type TextVariants = { color: 'light' | 'dark' };`

Is there a way to generate this type?

**Comment by daformat** [comment, marked as answer]
`RecipeVariants` includes `undefined`, here's what I do when I need something like this:

```ts
export type MaybeTextVariants = RecipeVariants<typeof textRecipe>;
export type TextVariants = Exclude<MaybeTextVariants, undefined>;
export type TextColor = TextVariants['color'];
```

I think you can go with just `Exclude<MaybeTextVariants, undefined>` in your case? The types shown in intellisense are confusing, but they are working as expected, hope this helps!

**Reply by laura-asimov** [reply]
@daformat Thank you!!
This is exactly what I was looking for, and for some reason it returns the string literals:

<img width="604" alt="image" src="https://github.com/vanilla-extract-css/vanilla-extract/assets/111299974/36d0fe28-2868-423e-be13-139a5a77e015">

**Reply by laura-asimov** [reply]
For some reason the types in intellisense revert back to being confusing when using type intersection:

<img width="604" alt="image" src="https://github.com/vanilla-extract-css/vanilla-extract/assets/111299974/f1d03558-d87a-45ce-b93e-a3b20c1f16d3">

**Reply by askoufis** [reply]
@laura-asimov Typescript will sometimes leave complex types un-evaluated until it actually needs to evaluate them. It can sometimes helps to wrap these types in a [prettify type](https://twitter.com/mattpocockuk/status/1622730173446557697) that forces typescript to expand the type completely. In some cases this doesn't always work though.

```ts
type Prettify<T> = {
  [K in keyof T]: T[K]
} & {}
